RecoTracker/LSTCore/src/alpaka/TrackCandidate.h

```diff
@@ -456,6 +528,216 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
+      int* sharedBestQuintupletIndex = alpaka::declareSharedVar<int[lst::kLogicalOtLayers], __COUNTER__>(acc);
+      int* sharedBestLayerSlotInQuintuplet = alpaka::declareSharedVar<int[lst::kLogicalOtLayers], __COUNTER__>(acc);
+
+      const unsigned int trackCandidateIndex = alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u];
+      const unsigned int nTrackCandidates = candsBase.nTrackCandidates();
+
+      if (trackCandidateIndex >= nTrackCandidates)
+        return;
+
+      // Initialize shared memory once per block
+      if (cms::alpakatools::once_per_block(acc)) {
+        for (int logicalLayerBin = 0; logicalLayerBin < lst::kLogicalOtLayers; ++logicalLayerBin) {
+          sharedBestScoreBits[logicalLayerBin] = bitCastFloatToInt(0.f);  // Stores the highest DNN score.
+          sharedBestQuintupletIndex[logicalLayerBin] = -1;
+          sharedBestLayerSlotInQuintuplet[logicalLayerBin] = -1;
+        }
+      }
+      alpaka::syncBlockThreads(acc);
+
+      const LSTObjType trackCandidateType = candsBase.trackCandidateType()[trackCandidateIndex];
+      if (!(trackCandidateType == LSTObjType::T5 || trackCandidateType == LSTObjType::pT5))
+        return;
+
+      // Base quintuplet (for T5: objectIndices[0], for pT5: objectIndices[1])
+      const unsigned int baseQuintupletIndex = (trackCandidateType == LSTObjType::T5)
+                                                   ? candsExtended.objectIndices()[trackCandidateIndex][0]
+                                                   : candsExtended.objectIndices()[trackCandidateIndex][1];
+
+      const float baseEta = __H2F(quintuplets.eta()[baseQuintupletIndex]);
+      const float basePhi = __H2F(quintuplets.phi()[baseQuintupletIndex]);
+      const uint8_t baseStartLogicalLayer = quintuplets.logicalLayers()[baseQuintupletIndex][0];
+
+      // Module range to scan
+      int lowerModuleBegin = 0;
+      int lowerModuleEnd = static_cast<int>(modules.nLowerModules());
+
+      // If starting at layer 1, restrict to the module of the second hit
+      if (baseStartLogicalLayer == 1) {
+        lowerModuleBegin = quintuplets.lowerModuleIndices()[baseQuintupletIndex][1];
+        lowerModuleEnd = lowerModuleBegin + 1;
+      }
+
+      // Flatten 2D thread indices within the block (Y, X) into one index
+      const int threadIndexFlat = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[1u] *
+                                      alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[2u] +
+                                  alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[2u];
+
+      const int blockDimFlat = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[1u] *
+                               alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[2u];
+
+      // Scan over lower modules
+      for (int lowerModuleIndex = lowerModuleBegin + threadIndexFlat; lowerModuleIndex < lowerModuleEnd;
```

Comment:
do we have any cases where two mergeable T5s would start more than one layer apart?
For the neighbor layer only, it's more efficient to use the linked modules from the module map (this may need an inward map; not sure we have it already, but should be easy to compute)